tooling/profiler/src/cli/execution_flamegraph_cmd.rs

```diff
@@ -0,0 +1,95 @@
+use std::path::{Path, PathBuf};
+
+use acir::circuit::OpcodeLocation;
+use acir::FieldElement;
+use clap::Args;
+use color_eyre::eyre::{self, Context};
+
+use crate::flamegraph::{FlamegraphGenerator, InfernoFlamegraphGenerator, Sample};
+use crate::fs::{read_inputs_from_file, read_program_from_file};
+use crate::opcode_formatter::AcirOrBrilligOpcode;
+use bn254_blackbox_solver::Bn254BlackBoxSolver;
+use nargo::ops::DefaultForeignCallExecutor;
+use noirc_abi::input_parser::Format;
+use noirc_artifacts::debug::DebugArtifact;
+
+#[derive(Debug, Clone, Args)]
+pub(crate) struct ExecutionFlamegraphCommand {
+    /// The path to the artifact JSON file
+    #[clap(long, short)]
+    artifact_path: PathBuf,
+
+    #[clap(long, short)]
+    prover_toml_path: PathBuf,
+
+    /// The output folder for the flamegraph svg files
+    #[clap(long, short)]
+    output: PathBuf,
+}
+
+pub(crate) fn run(args: ExecutionFlamegraphCommand) -> eyre::Result<()> {
+    run_with_generator(
+        &args.artifact_path,
+        &args.prover_toml_path,
+        &InfernoFlamegraphGenerator { count_name: "samples".to_string() },
+        &args.output,
+    )
+}
+
+fn run_with_generator(
+    artifact_path: &Path,
+    prover_toml_path: &Path,
+    flamegraph_generator: &impl FlamegraphGenerator,
+    output_path: &Path,
+) -> eyre::Result<()> {
+    let program =
+        read_program_from_file(artifact_path).context("Error reading program from file")?;
+
+    let (inputs_map, _) = read_inputs_from_file(prover_toml_path, Format::Toml, &program.abi)?;
+
+    let initial_witness = program.abi.encode(&inputs_map, None)?;
+
+    println!("Executing");
+    let (_, profiling_samples) = nargo::ops::execute_program_with_profiling(
+        &program.bytecode,
+        initial_witness,
+        &Bn254BlackBoxSolver,
+        &mut DefaultForeignCallExecutor::new(true, None, None, None),
+    )?;
+    println!("Executed");
+
+    let profiling_samples: Vec<Sample<FieldElement>> = profiling_samples
+        .into_iter()
+        .map(|sample| {
+            let call_stack = sample.call_stack;
+            let brillig_function_id = sample.brillig_function_id;
+            let last_entry = call_stack.last();
+            let opcode = brillig_function_id
```

Comment:
Can the `sample.opcode` already be `Some` here?

Reply:
`sample` is a `ProfilingSample` type which does not have an opcode field. `opcode` on this line is already an option as we are mapping `brillig_function_id` which is an option.